File: 0238-product-of-array-except-self/0238-product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    let len = nums.length
    
    let pre = [nums[0]]
    for (let i = 1; i < len; i++) {
        pre[i] = pre[i - 1] * nums[i]
    }
    console.log(pre)
    
    
    
    let post = []
    post[len - 1] = nums[len - 1]
    for (let i = len - 2; i >= 0; i--) {
        post[i] = post[i + 1] * nums[i]
    }
    console.log(post)
    
    
    let ans = []
    for (let i = 0; i < len; i++) {
        let prePart = (pre[i - 1] !== undefined) ? pre[i - 1] : 1    // be careful since 0 is considered falsy
        console.log(`pre: ${prePart}`)
        let postPart = (post[i + 1] !== undefined) ? post[i + 1] : 1 // be careful since 0 is considered falsy
        console.log(`post: ${postPart}`)

        ans[i] = prePart * postPart
    }
    console.log(ans)
    return ans
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # pre-fix, post-fix, then combine. that's O(3n) = O(n)
        
        # prefix
        prefix = []
        
        for i, n in enumerate(nums):
            if (i == 0):
                prefix.append(n)
            else:
                prefix.append(n * prefix[i-1])
        
        print(prefix)
              
              
        # postfix
        postfix = []
        
        for i, n in enumerate(reversed(nums)):
            if (i == 0):
                postfix.append(n)
            else:
                postfix.append(n * postfix[i-1])
        
        postfix.reverse()
        print(postfix)
        
        
        # put together
        ans = []
        
        for i, n in enumerate(nums):
            num1 = 1 if i-1 < 0 else prefix[i-1]
            num2 = 1 if i+1 > (len(nums) - 1) else postfix[i+1]
            ans.append(num1 * num2)
            
        
        return ans
            
            